cdb2api/cdb2api.c

```diff
@@ -3267,34 +3304,44 @@ static int cdb2_run_statement_typed_int(cdb2_hndl_tp *hndl, const char *sql,
     }
 
     if (type == RESPONSE_HEADER__SQL_RESPONSE_SSL) {
+#if WITH_SSL
         hndl->s_sslmode = PEER_SSL_REQUIRE;
-        try_ssl(hndl, hndl->sb, hndl->connected_host);
+        /* server wants us to use ssl so turn ssl on in same connection */
+        int resp = try_ssl(hndl, hndl->sb, hndl->connected_host);
+        if (resp != 0) {
+            PRINT_RETURN(-1);
+        }
+
         /* Decrement retry counter: It is not a real retry. */
         --retries_done;
+#else
+        PRINT_RETURN(-1);
+#endif
+        RETRY_QUERIES();
         goto retry_queries;
     }
 
     /* Dbinfo .. go to new node */
     if (type == RESPONSE_HEADER__DBINFO_RESPONSE) {
-        /* The master sent info about nodes that might be coherent. */
-        newsql_disconnect(hndl, hndl->sb, __LINE__);
-        CDB2DBINFORESPONSE *dbinfo_response = NULL;
-        dbinfo_response =
-            cdb2__dbinforesponse__unpack(NULL, len, hndl->first_buf);
-        parse_dbresponse(dbinfo_response, hndl->hosts, hndl->ports,
+        if (hndl->flags & CDB2_DIRECT_CPU) { 
+            /* direct cpu should not do anything with dbinfo */
+            PRINT_RETURN(-1);
```

Comment:
Okay. If the connected node becomes incoherent and sends back dbinfo response, under your logic, do we retry on next node or return an error to application (assuming user has defined multiple hosts)?